pkg/controller/direct/alloydb/cluster_controller.go

```diff
@@ -17,339 +17,644 @@ package alloydb
+	id, err := krm.NewClusterIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
 	}
 
-	projectID := obj.Spec.ProjectRef.External
-	if projectID == "" {
-		return nil, fmt.Errorf("cannot resolve project")
-	}
-	mapCtx := &direct.MapContext{
-		//	kube: kube,
-	}
-	desired := ClusterSpecToApi(mapCtx, &obj.Spec)
-	if mapCtx.Err() != nil {
-		return nil, mapCtx.Err()
-	}
-	// apply labels
-	desired.Labels = make(map[string]string)
-	desired.Labels["managed-by-cnrm"] = "true"
-	for k, v := range obj.Labels {
-		desired.Labels[k] = v
+	// Get alloydb GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
 	}
-
-	return &clusterAdapter{
-		resourceID: resourceID,
-		projectID:  projectID,
-		location:   location,
-		desired:    desired,
-		client:     client,
+	return &ClusterAdapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+		reader:    reader,
 	}, nil
 }
 
-func (m *clusterModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+func (m *modelCluster) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
 	return nil, nil
 }
 
-// adapter implements the Adapter interface.
-var _ directbase.Adapter = &clusterAdapter{}
+type ClusterAdapter struct {
+	id        *krm.ClusterIdentity
+	gcpClient *gcp.AlloyDBAdminClient
+	desired   *krm.AlloyDBCluster
+	actual    *alloydbpb.Cluster
+	reader    client.Reader
+}
 
-// Find implements the Adapter interface.
-func (a *clusterAdapter) Find(ctx context.Context) (bool, error) {
-	if a.resourceID == "" {
-		return false, nil
-	}
-	cluster, err := a.client.Projects.Locations.Clusters.Get(a.fullyQualifiedName()).Context(ctx).Do()
+var _ directbase.Adapter = &ClusterAdapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return false means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *ClusterAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting Cluster", "name", a.id)
+
+	req := &alloydbpb.GetClusterRequest{Name: a.id.String()}
+	clusterpb, err := a.gcpClient.GetCluster(ctx, req)
 	if err != nil {
 		if direct.IsNotFound(err) {
 			return false, nil
 		}
-		return false, err
+		return false, fmt.Errorf("getting Cluster %q: %w", a.id, err)
 	}
 
-	a.actual = cluster
-
+	a.actual = clusterpb
 	return true, nil
 }
 
-// Delete implements the Adapter interface.
-func (a *clusterAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
-	// Already deleted
-	if a.resourceID == "" {
-		return false, nil
+// TODO: Scenario test cases: both networkConfig.networkRef and networkRef set; none set.
+func (a *ClusterAdapter) resolveNetworkRef(ctx context.Context) error {
+	obj := a.desired
+	if obj.Spec.NetworkRef == nil && obj.Spec.NetworkConfig == nil {
```

Comment:
Have we considered adding CEL rules for these checks, which allows us to shift the validation earlier in the process.

Reply:
Yeah probably not. Even if it can be done, the logic is probably more readable if implemented here.